### PR TITLE
fix(failsafe): Inform user about every failsafe triggered

### PR DESCRIPTION
**Pull request by Frigiii**
### Solved Problem
If a warn-level failsafe triggers (like motor failure) while in termination, the user is notified about the termination. This behavior spams the user with meaningless notifications, if the failsafe switches continuously.

### Solution
Inform the user about every failsafe triggered. If the failsafe is not highest severity, we inform the user with a warn-level failsafe notification.

### Changelog Entry
For release notes:
```
Bugfix: Inform user about every failsafe triggered.
```

### CODE REVIEW

File: src/modules/commander/failsafe/framework.cpp [hunk 1]
```diff
@@ -89,8 +89,8 @@ uint8_t FailsafeBase::update(const hrt_abstime &time_us, const State &state, boo
 	updateStartDelay(time_us - _last_update, action_state.delayed_action != Action::None);
 	updateFailsafeDeferState(time_us, action_state.failsafe_deferred);
 
-	// Notify user if the action is worse than before, or a new action got added
-	if (action_state.action > _selected_action || (action_state.action != Action::None && _notification_required)) {
+	// Notify user if the action is worse than before, or a new warn action got added
+	if (action_state.action > _selected_action || (action_state.action == Action::Warn && _notification_required)) {
```
Comment (Frigiii) [inline src/modules/commander/failsafe/framework.cpp]: Alternatively, we could also only trigger the notification, while not in termination.
```suggestion
	// Notify user if the action is worse than before, or a new action got added
	if (action_state.action > _selected_action || (action_state.action != Action::None && action_state.action != Action::Terminate && _notification_required)) {
```

**Review by MaEtUgR — COMMENTED** [review]
I took a closer look and got to the following conclusion:
- It's handling for multiple concurrent failsafes.
- If the new failsafe is more severe than the existing one the handling works as expected.
- If the new failsafe is less severe then the existing one then the `_notification_required` logic previously only passed through new warnings but the notification to the user was wrong. E.g. reinforming about termiantion when a different warning occurred after that.
- @Frigiii 's commit restricts passing through warning to if the previous failsafe was a warning (nothing more severe).

My suggestion is:
- For any new less severe failsafe, warn the user.
- Warn him about the new cause (saved instead of bool `_notification_required`) and as if the action was a warning because it wasn't more severe and there's no actual change of mode.

I tested this in SIH with the following case:
- multirotor: `make px4_sitl sihsim_hex`
- Flying in position mode really high up
- RC loss -> RTL
- Battery draining during RTL which low -> warning, critical -> RTL (not more severe), emergency -> land (more severe)

The messageing in QGC is:
```
[15:13:28.763 ] Info: Manual control lost
[15:13:28.776 ] Critical: Manual control loss: switching to RTL in 5 seconds
[15:13:33.849 ] Critical: Manual control loss: switching to RTL
[15:13:33.853 ] Info: RTL: start return at 865 m (376 m above destination)
[15:14:19.168 ] Warning: Low battery level, return advised
[15:14:24.097 ] Critical: Critical battery level, land now
[15:14:25.307 ] EMERGENCY: Dangerously low battery! System shut down failed
Cannot shut down, most likely the system does not support it.
[15:14:25.308 ] Critical: Emergency battery level: switching to Land
```

Except for the shutdown message this looks good but I'm unsure if that' because of my suggested change or not, need to find an obiouds test case 👀

**Comment by MaEtUgR** [comment]
It doesn't work as I'd expect in my latest test:
I flew in Position mode, unplugged the (virtual) joystick to have RC loss which by default leads to Return mode and then during that RTL set `COM_PARACHUTE` to `2` which by itself immediately triggers RTL because the sim has no parachute but in the test order it doesn't output any warning like I'd have expected. I'm unsure if another assumption doesn't hold and if it only works for warning failsafes because of that.

```
[15:14:25.308 ] Critical: Emergency battery level: switching to Land
[15:14:25.307 ] EMERGENCY: Dangerously low battery! System shut down failed
Cannot shut down, most likely the system does not support it.
[15:14:24.097 ] Critical: Critical battery level, land now
[15:14:19.168 ] Warning: Low battery level, return advised
[15:13:33.853 ] Info: RTL: start return at 865 m (376 m above destination)
[15:13:33.849 ] Critical: Manual control loss: switching to RTL
[15:13:28.776 ] Critical: Manual control loss: switching to RTL in 5 seconds
[15:13:28.763 ] Info: Manual control lost
[15:12:02.149 ] Info: Pilot took over using sticks
[15:09:52.580 ] Info: Takeoff detected
[15:09:50.846 ] Info: Armed by external command
```

**Comment by Frigiii** [comment]
Test case: 
- Standard VTOL in simulation
- QGC Groundstation
- Fly around geofence exclusion zone, in altitude mode, and then into it -> RTL triggers
- While flying through geofence disconnect the controller, which is also configured to trigger RTL.

I think that looks good for this test case.

```
[10:10:34.490 ] Warning: Failsafe warning: Manual control loss
No action is triggered.
[10:10:34.479 ] Info: Manual control lost
[10:10:34.476 ] Critical: Geofence: approaching or outside geofence
[10:10:16.889 ] Info: RTL: start return at 61 m (60 m above destination)
[10:10:16.888 ] Critical: Failsafe activated: switching to RTL
Geofence violation: approaching or outside geofence
[10:10:11.886 ] Critical: Failsafe activated: switching to RTL in 5 seconds
Geofence violation: approaching or outside geofence
[10:10:11.869 ] Critical: Geofence: approaching or outside geofence
[10:08:48.271 ] Info: Takeoff detected
[10:08:48.181 ] Info: Pilot took over using sticks
[10:08:46.430 ] Info: Armed by external command
```

**Comment by Frigiii** [comment]
Test case:

- Quadcopter hardware
- QGC Groundstation
- Disarm, spin up motors
- Disconnect controller, which is configured to trigger terminate
- An edge case repeatedly sets and unsets the ESC's as offline

I think that looks good for this test case.


```
No action is triggered.
[10:56:24.819 ] Warning: Failsafe warning:
- ESC 2 offline
- No manual control input
- Flight termination active

No action is triggered.
[10:56:24.639 ] Warning: Failsafe warning:
- ESC 2 offline
- No manual control input
- Flight termination active

No action is triggered.
[10:56:24.420 ] Warning: Failsafe warning:
- ESC 2 offline
- No manual control input
- Flight termination active

No action is triggered.
[10:56:22.855 ] Warning: Failsafe warning:
- ESC 1 offline
- ESC 2 offline
- ESC 3 offline
- No manual control input
- Flight termination active

No action is triggered.
[10:56:22.553 ] Critical: Manual control loss: switching to terminate
[10:56:22.552 ] Info: Manual control lost
[10:56:20.027 ] Info: Takeoff detected
[10:56:16.550 ] Info: Armed by external command
```

File: src/modules/commander/failsafe/framework.cpp [hunk 2]
```diff
@@ -361,9 +364,7 @@ bool FailsafeBase::checkFailsafe(int caller_id, bool last_state_failure, bool cu
 					}
 				}
 
-				if (options.action == Action::Warn) {
-					_notification_required = true;
-				}
+				_pending_notification_cause = options.cause;
```
Comment (bkueng) [inline src/modules/commander/failsafe/framework.cpp]: You probably have to check if `options.action != Action::None`

Comment (Frigiii) [reply src/modules/commander/failsafe/framework.cpp]: Good catch. This will prevent notification on disabled failsafe.

**Review by bkueng — APPROVED** [review]
Looks good from my side
